build-monitor-plugin/src/main/java/com/smartcodeltd/jenkinsci/plugins/buildmonitor/viewmodel/BuildView.java

```diff
@@ -201,4 +224,32 @@ private BuildView(Run<?, ?> build, RelativeLocation parentJobLocation, Date syst
 
         return users;
     }
+
+    private static class StageTraversal {
+
+        private final Queue<FlowNode> nodesToAccess;
+        private final List<String> stages;
+
+        public StageTraversal() {
+            this.nodesToAccess = new LinkedList<FlowNode>();
+            this.stages = new ArrayList<String>();
+        }
+
+        public void start(List<FlowNode> nodes) {
+            nodesToAccess.addAll(nodes);
+            while (!nodesToAccess.isEmpty()) {
+                FlowNode flowNode = nodesToAccess.remove();
+                if (flowNode instanceof StepStartNode
+                        && ((StepStartNode) flowNode).getDescriptor().isSubTypeOf(StageStep.class)) {
+                    stages.add(flowNode.getDisplayName());
+                } else {
+                    nodesToAccess.addAll(flowNode.getParents());
```

Comment:
This also potentially causes an NPE. A null-check for flowNode && flowNode.getParents() is probably necessary.